Use method entries instead of primary ranges for method DIE generation

zakkak opened the pull request:
As of https://github.com/oracle/graal/pull/3304 we associate every range
with the `MethodEntry` of the method it was compiled from, and as a
result we know that a `MethodEntry` exists for each compiled method in
the image. Actually we create `MethodEntry`s even for methods that are
not compiled but are part of `ClassEntry` that describes a Class
reachable in the image.

For every range we generate a DIE (Debugging Information Entry) which
references another DIE that holds the specification of the method that
this range was generated from. So for every range we need a method DIE
containing information for the corresponding method.

Till now we have been generating these DIEs by traversing the primary
ranges of each ClassEntry. There are, however, subranges that might be
compiled from a method that is not referenced by a primary range, e.g.,
inlined methods.

This patch ensures that every method that is compiled in the native
image (including inlined ones) will get a method DIE.

zakkak commented:
@adinn this is complementary to #3304 and ensures we have a method DIE for every method referenced by a code range.
Can you please review?

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfInfoSectionImpl.java:
```diff
@@ -654,43 +655,47 @@ private int writeField(DebugContext context, StructureTypeEntry entry, FieldEntr
+        String name = uniqueDebugString(method.methodName());
         log(context, "  [0x%08x]     name 0x%x (%s)", pos, debugStringIndex(name), name);
         pos = writeAttrStrp(name, buffer, pos);
         int fileIdx = classEntry.localFilesIdx();
```
The next few lines need rebasing against head. I suggest this

        FileEntry fileEntry = method.getFileEntry();
        int fileIdx;
        if (fileEntry != null) {
            fileIdx = classEntry.localFilesIdx(fileEntry);
        } else {
            fileEntry = classEntry.getFileEntry();
            fileIdx = classEntry.localFilesIdx();
        }
        log(context, "  [0x%08x]     file 0x%x (%s)", pos, fileIdx, fileEntry.getFullName());

zakkak replied:
I prefer:
```
        FileEntry fileEntry = method.getFileEntry();
        if (fileEntry == null) {
            fileEntry = classEntry.getFileEntry();
        }
        int fileIdx = classEntry.localFilesIdx(fileEntry);
```
If there isn't anything wrong with it.

adinn replied:
Sure, it comes to the same thing.

adinn reviewed (changes requested):
This looks fine modulo a few small changes to make it clearer what is going on and a bit of finessing the rebase against the change pending in head

zakkak commented:
@adinn I have updated the PR to resolve your comments, please re-review :)

adinn reviewed (approved):
Thanks, looks good.

zakkak commented:
It looks like something broke with the rebase. Looking into it!

zakkak commented:
It turns out this will have to wait for #3421 to get merged first.

zakkak commented:
Now that #3421 this PR is ready for (re)review.

@adinn if I am not mistaken the new changes that you need to check are: 

* https://github.com/oracle/graal/pull/3419/commits/0b665d320149f292a537ebc43bf6171cfff5a212
* https://github.com/oracle/graal/pull/3419/commits/e205cb000de10793e0f9e10d36d2a57caeb8e538
* https://github.com/oracle/graal/pull/3419/commits/981590d3a4866b3fd888f684e24653ee1b0b972e
* https://github.com/oracle/graal/pull/3419/commits/a0cf3be29aaa5b635aa2ace659c39f976f4f56e0

zakkak commented:
CI failures seem unrelated to the PR.

adinn reviewed (approved):
PR is still reviewed after latest changes

adinn commented:
@olpaw This next patch is now ready for review. It looks good to me.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:
```diff
@@ -99,7 +100,8 @@ public ClassEntry(String className, FileEntry fileEntry, int size) {
         // it sorted,
-        // so a LinkedList seems more appropriate than an ArrayList. (see getMethodEntry)
+        // so a LinkedList seems more appropriate than an ArrayList.
+        // (see ensureMethodEntryForDebugRangeInfo)
```
How is using a `LinkedList` more appropriate than an `ArrayList`?

It's the opposite. While `ArrayList` can use an optimized sort (`java.util.ArrayList#sort` performing an `Arrays.sort` directly on the underlying `elementData` array, `LinkedList` has to use the fallback implementation that first turns the elements into a temporary array that then gets sorted and then gets written back into the `LinkedList` via ListIterator.

olpaw replied:
Please grep through the whole debuginfo sourcebase and replace all occurrences of LinkedList with ArrayList. If you find a place where you really have to use LinkedList for a good reason please let me know.

zakkak replied:
I fount `LinkedList` more appropriate because despite the less efficient sort, it allows us to perform a sorted insertion in https://github.com/oracle/graal/pull/3419/files/a0cf3be29aaa5b635aa2ace659c39f976f4f56e0#diff-b7692d10d2dc550e4bbf155d2f7f142dbdaca4aca41c2677d41b193149c168a4R366 without having to shift elements.
Now, given that adding more methods to the `ClassEntry` during code processing and thus needing sorted insertions appears to be rare (empirically), the cost of shifting the elements might indeed be not significant, so I think we could go with `ArrayList` as well.

zakkak replied:
Resolved in https://github.com/oracle/graal/pull/3419/commits/6bd40df1bf617ed2532fd9136d8daf67448f4bac

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DebugInfoBase.java:
```diff
@@ -390,10 +389,12 @@ private FileEntry addFileEntry(String fileName, Path filePath, Path cachePath) {
     }
 
-    protected FileEntry ensureFileEntry(String fileName, Path filePath, Path cachePath) {
+    protected FileEntry ensureFileEntry(DebugFileInfo debugFileInfo) {
```
:+1:

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java:
```diff
@@ -26,18 +26,27 @@
     final TypeEntry[] paramTypes;
     final String[] paramNames;
     final boolean isDeoptTarget;
+    boolean isInRange;
```
Any reason why `isInRange` should not be also final?

For final primitive fields it often makes sense to just have them public and not have an accessor method. I.e. make `isDeoptTarget` and `isInRange` public final and remove the accessor method boilerplate.

zakkak replied:
Resolved in https://github.com/oracle/graal/pull/3419/commits/d73b4e74d453840fe0c20c018045702ce880aeab

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java:
```diff
@@ -74,33 +87,41 @@ public String getParamName(int idx) {
+            return;
         }
-        return 0;
+        isInRange = true;
```
Ah I see. You can have it final because of this.

olpaw reviewed (changes requested):
Looks good overall. Small changes requested as described in the other comments.

zakkak commented:
Thank you Paul.

olpaw commented:
When running on Windows uur CI found the following issue with this PR when building the gu tool:
```
[08:58:12] Archiving GRAALVM_DFECB19112_JAVA16_STAGE1 [duration: 00:11.333143]
[08:58:13] Building gu.exe.image... [dependency INSTALLER updated]
[gu:22748]    classlist:  12,421.83 ms,  0.96 GB
[gu:22748]        (cap):   2,295.61 ms,  0.94 GB
[gu:22748]        setup:   7,654.80 ms,  0.94 GB
[gu:22748]     (clinit):     775.04 ms,  3.62 GB
[gu:22748]   (typeflow):  24,525.57 ms,  3.62 GB
[gu:22748]    (objects):  31,850.16 ms,  3.62 GB
[gu:22748]   (features):   2,232.18 ms,  3.62 GB
[gu:22748]     analysis:  61,130.92 ms,  3.62 GB
[gu:22748]     universe:   3,532.48 ms,  3.64 GB
[gu:22748]      (parse):   3,149.61 ms,  3.64 GB
[gu:22748]     (inline):  10,165.71 ms,  5.27 GB
[gu:22748]    (compile):  79,771.56 ms,  6.17 GB
[gu:22748]      compile:  97,142.31 ms,  6.17 GB
[gu:22748]      dbginfo: 112,748.82 ms,  6.67 GB
[gu:22748]        image: 120,449.73 ms,  6.67 GB
[gu:22748]        write:      33.05 ms,  6.67 GB
Fatal error:com.oracle.svm.core.util.VMError$HostedError: java.lang.NullPointerException: Cannot invoke "com.oracle.objectfile.debugentry.FileEntry.getFileName()" because "this.methodEntry.fileEntry" is null
	at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:72)
	at com.oracle.svm.hosted.image.NativeImage.write(NativeImage.java:172)
	at com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:409)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:709)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:494)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1414)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:295)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.NullPointerException: Cannot invoke "com.oracle.objectfile.debugentry.FileEntry.getFileName()" because "this.methodEntry.fileEntry" is null
	at com.oracle.objectfile.debugentry.Range.getFileName(Range.java:164)
	at com.oracle.objectfile.pecoff.cv.CVLineRecordBuilder.build(CVLineRecordBuilder.java:65)
	at com.oracle.objectfile.pecoff.cv.CVSymbolSubsectionBuilder.addLineNumberRecords(CVSymbolSubsectionBuilder.java:148)
	at com.oracle.objectfile.pecoff.cv.CVSymbolSubsectionBuilder.build(CVSymbolSubsectionBuilder.java:116)
	at com.oracle.objectfile.pecoff.cv.CVSymbolSubsectionBuilder.build(CVSymbolSubsectionBuilder.java:76)
	at com.oracle.objectfile.pecoff.cv.CVSymbolSubsectionBuilder.build(CVSymbolSubsectionBuilder.java:63)
	at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.addFunctionRecords(CVSymbolSectionImpl.java:126)
	at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.addRecords(CVSymbolSectionImpl.java:108)
	at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.createContent(CVSymbolSectionImpl.java:72)
	at com.oracle.objectfile.ObjectFile.debugContext(ObjectFile.java:1845)
	at com.oracle.objectfile.pecoff.cv.CVSectionImpl.getOrDecideContent(CVSectionImpl.java:81)
	at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.getOrDecideContent(CVSymbolSectionImpl.java:41)
	at com.oracle.objectfile.pecoff.PECoffUserDefinedSection.getOrDecideContent(PECoffUserDefinedSection.java:104)
	at com.oracle.objectfile.ObjectFile.bake(ObjectFile.java:1630)
	at com.oracle.objectfile.ObjectFile.write(ObjectFile.java:1266)
	at com.oracle.svm.hosted.image.NativeImage.lambda$write$0(NativeImage.java:168)
	at com.oracle.objectfile.ObjectFile.withDebugContext(ObjectFile.java:1826)
	at com.oracle.svm.hosted.image.NativeImage.write(NativeImage.java:167)
	... 9 more
```

olpaw commented:
What can also be seen in https://github.com/oracle/graal/pull/3419#issuecomment-855753798 is that the `dbginfo` phase absolutely dominates image build time. This is something that needs to be fixed. `dbginfo` generation should only take a fraction of the total image build time.

adinn commented:
Hi Foivos.

> Reverting e205cb0 resolves the issue. e205cb0 was initially applied because I wanted a different MethodEntry for method substitutions, but that's not the case any more.

Yes, the old code looks good to me. However ...

In MethodEntry.compareTo I wonder if it would be better to avoid the code that splits and trims the signature argument and compares each of its elements against the param types. If a MethodEntry created and saved the relevant signature string in its constructor then we could just do a one off string compare. Could you try working that in to the fix?

zakkak commented:
Hi @adinn I gave the following a go:
```diff
diff --git a/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java
index c9c004cf23b..91c38a07824 100644
--- a/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java
+++ b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java
@@ -28,6 +28,9 @@ package com.oracle.objectfile.debugentry;
 
 import com.oracle.objectfile.debuginfo.DebugInfoProvider.DebugRangeInfo;
 
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
 public class MethodEntry extends MemberEntry implements Comparable<MethodEntry> {
     final TypeEntry[] paramTypes;
     final String[] paramNames;
@@ -35,6 +38,7 @@ public class MethodEntry extends MemberEntry implements Comparable<MethodEntry>
     boolean isInRange;
 
     final String symbolName;
+    private String signature;
 
     public MethodEntry(FileEntry fileEntry, String symbolName, String methodName, ClassEntry ownerType,
                     TypeEntry valueType, TypeEntry[] paramTypes, String[] paramNames, int modifiers,
@@ -43,6 +47,7 @@ public class MethodEntry extends MemberEntry implements Comparable<MethodEntry>
         assert ((paramTypes == null && paramNames == null) ||
                         (paramTypes != null && paramNames != null && paramTypes.length == paramNames.length));
         this.paramTypes = paramTypes;
+        this.signature = Arrays.stream(paramTypes).map(type -> type.typeName).collect(Collectors.joining(","));
         this.paramNames = paramNames;
         this.isDeoptTarget = isDeoptTarget;
         this.isInRange = isInRange;
@@ -129,24 +134,7 @@ public class MethodEntry extends MemberEntry implements Comparable<MethodEntry>
         if (typeComparison != 0) {
             return typeComparison;
         }
-        String[] paramTypeNames = paramSignature.split((","));
-        int length;
-        if (paramSignature.trim().length() == 0) {
-            length = 0;
-        } else {
-            length = paramTypeNames.length;
-        }
-        int paramCountComparison = getParamCount() - length;
-        if (paramCountComparison != 0) {
-            return paramCountComparison;
-        }
-        for (int i = 0; i < getParamCount(); i++) {
-            int paraComparison = getParamTypeName(i).compareTo(paramTypeNames[i].trim());
-            if (paraComparison != 0) {
-                return paraComparison;
-            }
-        }
-        return 0;
+        return signature.compareTo(paramSignature);
     }
 
     @Override
@@ -160,16 +148,6 @@ public class MethodEntry extends MemberEntry implements Comparable<MethodEntry>
         if (typeComparison != 0) {
             return typeComparison;
         }
-        int paramCountComparison = getParamCount() - other.getParamCount();
-        if (paramCountComparison != 0) {
-            return paramCountComparison;
-        }
-        for (int i = 0; i < getParamCount(); i++) {
-            int paramComparison = getParamTypeName(i).compareTo(other.getParamTypeName(i));
-            if (paramComparison != 0) {
-                return paramComparison;
-            }
-        }
-        return 0;
+        return signature.compareTo(other.signature);
     }
 }
```
But that seems to increase compilation time significantly.

I even tried with:
```diff
diff --git a/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java
index c9c004cf23b..064d4f4fba6 100644
--- a/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java
+++ b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/MethodEntry.java
@@ -28,6 +28,9 @@ package com.oracle.objectfile.debugentry;
 
 import com.oracle.objectfile.debuginfo.DebugInfoProvider.DebugRangeInfo;
 
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
 public class MethodEntry extends MemberEntry implements Comparable<MethodEntry> {
     final TypeEntry[] paramTypes;
     final String[] paramNames;
@@ -35,6 +38,7 @@ public class MethodEntry extends MemberEntry implements Comparable<MethodEntry>
     boolean isInRange;
 
     final String symbolName;
+    private String signature;
 
     public MethodEntry(FileEntry fileEntry, String symbolName, String methodName, ClassEntry ownerType,
                     TypeEntry valueType, TypeEntry[] paramTypes, String[] paramNames, int modifiers,
@@ -120,6 +124,13 @@ public class MethodEntry extends MemberEntry implements Comparable<MethodEntry>
         return symbolName;
     }
 
+    private String getSignature() {
+        if (signature == null) {
+            signature = Arrays.stream(paramTypes).map(type -> type.typeName).collect(Collectors.joining(","));
+        }
+        return signature;
+    }
+
     public int compareTo(String methodName, String paramSignature, String returnTypeName) {
         int nameComparison = memberName.compareTo(methodName);
         if (nameComparison != 0) {
@@ -129,24 +140,7 @@ public class MethodEntry extends MemberEntry implements Comparable<MethodEntry>
         if (typeComparison != 0) {
             return typeComparison;
         }
-        String[] paramTypeNames = paramSignature.split((","));
-        int length;
-        if (paramSignature.trim().length() == 0) {
-            length = 0;
-        } else {
-            length = paramTypeNames.length;
-        }
-        int paramCountComparison = getParamCount() - length;
-        if (paramCountComparison != 0) {
-            return paramCountComparison;
-        }
-        for (int i = 0; i < getParamCount(); i++) {
-            int paraComparison = getParamTypeName(i).compareTo(paramTypeNames[i].trim());
-            if (paraComparison != 0) {
-                return paraComparison;
-            }
-        }
-        return 0;
+        return getSignature().compareTo(paramSignature);
     }
 
     @Override
@@ -160,16 +154,6 @@ public class MethodEntry extends MemberEntry implements Comparable<MethodEntry>
         if (typeComparison != 0) {
             return typeComparison;
         }
-        int paramCountComparison = getParamCount() - other.getParamCount();
-        if (paramCountComparison != 0) {
-            return paramCountComparison;
-        }
-        for (int i = 0; i < getParamCount(); i++) {
-            int paramComparison = getParamTypeName(i).compareTo(other.getParamTypeName(i));
-            if (paramComparison != 0) {
-                return paramComparison;
-            }
-        }
-        return 0;
+        return getSignature().compareTo(other.getSignature());
     }
 }
```
to avoid keeping signatures for method entries that we never reach, but it didn'e help either.

If you insist on goind that way I will need to do some profiling to see what's causing that overhead (unless it's something obvious I am missing and you can point me to it).

adinn commented:
Well, I am not surprised that the followng slows things down:

        this.signature = Arrays.stream(paramTypes).map(type -> type.typeName).collect(Collectors.joining(","));

Using a stream and a collect predicate is going to involve settng up a whole load of apparatus and processing overhead that a simple for loop with dedicated String appender object will probably do much faster.

How much does this code slow things down:

    StringBuilder builder = new  StringBuilder();
    String separator = "";
    builder.append("(");
    for (int i = 0; i < paramTypes.length; i++) {
      builder.append(separator);
      builder.append(paramTypes[i];
      separator = ",";
    }
    builder.append(")");
    this.signature = builder.toString();

If the above code also ends up adding more work then forget about precomputing the signature.

zakkak commented:
So it turns out that what was actually slowing things down was that I was not adding spaces between the parameters, so I was essentially creating new method entries (and performing insertion sort) for all the methods we need all over again.

```diff
- java.util.regex.Matcher, int, java.lang.CharSequence
+ java.util.regex.Matcher,int,java.lang.CharSequence
```

After fixing this the _string-builder vs stream vs string-split_ difference seems to be negligible.

Additionally, generating the signatures on demand and not in the constructor seems beneficial as it results in producing ~10900 signatures for the ~16600 `MethodEntry`s that we generate for the `hello.Hello` example we use in debuginfo testing.

zakkak commented:
> @zakkak when running on Windows our CI found the following issue with this PR when building the `gu` tool on Java 16

https://github.com/oracle/graal/pull/3419/commits/b9d4679849b5dc4faa1826ddc7f3c1eae1c1dfe8 should fix this. @stooke can you please give this a go with:
```
cd graal/vm
mx --dy /substratevm \
   --native-images=gu \
   --extra-image-builder-argument=-g \
   --extra-image-builder-argument=-J-ea \
   --extra-image-builder-argument=-J-esa \
   build
```
?

Thanks

stooke commented:
> > @zakkak when running on Windows our CI found the following issue with this PR when building the `gu` tool on Java 16
> 
> [12beae4](https://github.com/oracle/graal/commit/12beae445a51f6c2070fad83b67a4ea19af90528) should fix this. @stooke can you please give this a go with:
> 
> ```
> cd graal/vm
> mx --dy /substratevm \
>    --native-images=gu \
>    --extra-image-builder-argument=-g \
>    --extra-image-builder-argument=-J-ea \
>    --extra-image-builder-argument=-J-esa \
>    build
> ```
> 
> ?
> 
> Thanks

This completes successfully on my development VM.

zakkak commented:
> This completes successfully on my development VM.

Great, thank you Simon.

@olpaw both the NPE and the performance regression should now be fixed. Can you please have a look?
If all is good, please let me know before pushing it internally so that I can squash the revert and the fixup.

olpaw commented:
> @olpaw both the NPE and the performance regression should now be fixed. Can you please have a look?
If all is good, please let me know before pushing it internally so that I can squash the revert and the fixup.

Hi @zakkak I will re-run on our gates now to see if the debuginfo generation time regression and the NPE are now gone.

olpaw commented:
The debuginfo generation time regression is gone now but I can still observe NPEs on Windows gates. E.g.
```
[jvmcicompiler:8436]        (cap):   2,842.74 ms,  0.96 GB
[jvmcicompiler:8436]        setup:   6,989.66 ms,  0.96 GB
[jvmcicompiler:8436]     (clinit):     853.17 ms,  3.24 GB
[jvmcicompiler:8436]   (typeflow):  26,463.87 ms,  3.24 GB
[jvmcicompiler:8436]    (objects):  19,998.79 ms,  3.24 GB
[jvmcicompiler:8436]   (features):   1,705.22 ms,  3.24 GB
[jvmcicompiler:8436]     analysis:  51,013.47 ms,  3.24 GB
[jvmcicompiler:8436]     universe:   2,480.12 ms,  3.24 GB
0 method(s) included for runtime compilation
[jvmcicompiler:8436]      (parse):  11,047.72 ms,  3.80 GB
[jvmcicompiler:8436]     (inline):  12,553.71 ms,  5.76 GB
[jvmcicompiler:8436]    (compile):  84,399.77 ms,  5.95 GB
[jvmcicompiler:8436]      compile: 112,418.32 ms,  5.95 GB
[jvmcicompiler:8436]      dbginfo:  18,382.53 ms,  6.02 GB
[jvmcicompiler:8436]        image:  26,457.70 ms,  6.02 GB
[jvmcicompiler:8436]        write:   1,181.04 ms,  6.02 GB
[jvmcicompiler:8436]      [total]: 208,494.22 ms,  6.02 GB
# Printing build artifacts to: C:\buildbot\worker\e\graal\sdk\mxbuild\windows-amd64\jvmcicompiler.dll.image\jvmcicompiler.build_artifacts.txt
Fatal error:com.oracle.svm.core.util.VMError$HostedError: java.lang.NullPointerException
	at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:72)
	at com.oracle.svm.hosted.image.NativeImage.write(NativeImage.java:172)
	at com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:409)
	at com.oracle.svm.hosted.image.SharedLibraryImageViaCC.write(SharedLibraryImageViaCC.java:56)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:701)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:497)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.NullPointerException
	at com.oracle.objectfile.pecoff.PECoffSymtab.indexOf(PECoffSymtab.java:302)
	at com.oracle.objectfile.pecoff.PECoffRelocationTable.getNativeReloctab(PECoffRelocationTable.java:206)
	at com.oracle.objectfile.pecoff.PECoffRelocationTable.getOrDecideContent(PECoffRelocationTable.java:240)
	at com.oracle.objectfile.ObjectFile.bake(ObjectFile.java:1630)
	at com.oracle.objectfile.ObjectFile.write(ObjectFile.java:1266)
	at com.oracle.svm.hosted.image.NativeImage.lambda$write$0(NativeImage.java:168)
	at com.oracle.objectfile.ObjectFile.withDebugContext(ObjectFile.java:1826)
	at com.oracle.svm.hosted.image.NativeImage.write(NativeImage.java:167)
	... 10 more
```

zakkak commented:
Hmmm, @stooke can you please have a look?

@olpaw what would be the command to reproduce this?
Would
```
cd graal/vm
mx --dy /substratevm \
   --native-images=jvmcicompiler \
   --extra-image-builder-argument=-g \
   --extra-image-builder-argument=-J-ea \
   --extra-image-builder-argument=-J-esa \
   build
```
do the trick?

olpaw commented:
Usually the fastest way to reproduce/customize a specific image build is via
```
cd mandrel/substratevm
mx build
mx native-image --macro:jvmcicompiler-library -g
```
but unfortunately that doesn't work here since this is an EE gate that's failing.

@stooke do you need more info on this or is the stacktrace in https://github.com/oracle/graal/pull/3419#issuecomment-860659784 sufficient for you to identify the cause of the NPE?

stooke commented:
> Usually the fastest way to reproduce/customize a specific image build is via
> 
> ```
> cd mandrel/substratevm
> mx build
> mx native-image --macro:jvmcicompiler-library -g
> ```
> 
> but unfortunately that doesn't work here since this is an EE gate that's failing.
> 
> @stooke do you need more info on this or is the stacktrace in [#3419 (comment)](https://github.com/oracle/graal/pull/3419#issuecomment-860659784) sufficient for you to identify the cause of the NPE?

It's difficult to do if I can't reproduce it (I haven't yet had time to try; I was off yesterday and have a meeting first thing this morning).  It appears as if the relocation table is being written out before the symbol table.  I'll poke around, and if there are any pointers to reproduce it in the CE version, that'd be great.

olpaw commented:
> appears as if the relocation table is being written out before the symbol table. I'll poke around, and if there are any pointers to reproduce it in the CE version, that'd be great.

So far I found out that i can only reproduce it if build with `mx --strip-jars` so I think it only shows up if we have jar-files that have been stripped with proguard. Unfortunately I'm still not able to reproduce on CE.

The place where we issue occurs `PECoffRelocationTable.getNativeReloctab(PECoffRelocationTable.java:206` I see
```java
        for (PECoffSection s : getOwner().getPECoffSections()) {
            Map<Entry, Entry> entries = (Map<Entry, Entry>) s.getRelocEntries();
            if (entries != null) {
                for (Entry ent : entries.keySet()) {
                    long offset = ent.getOffset();
                    int sectionID = ent.section == null ? 0 : s.getSectionID();
                    relocTabStruct.createRelocationEntry(sectionID, (int) offset, syms.indexOf(ent.sym), (int) ent.t.toLong());
                }
            }
        }
```
here we have an `ent` in the innermost for where `ent.sym` is `null`. I will let you know if I find out more.

stooke commented:
> > appears as if the relocation table is being written out before the symbol table. I'll poke around, and if there are any pointers to reproduce it in the CE version, that'd be great.
> 
> So far I found out that i can only reproduce it if build with `mx --strip-jars` so I think it only shows up if we have jar-files that have been stripped with proguard. Unfortunately I'm still not able to reproduce on CE.
> 
> The place where we issue occurs `PECoffRelocationTable.getNativeReloctab(PECoffRelocationTable.java:206` I see
> 
> ```java
>         for (PECoffSection s : getOwner().getPECoffSections()) {
>             Map<Entry, Entry> entries = (Map<Entry, Entry>) s.getRelocEntries();
>             if (entries != null) {
>                 for (Entry ent : entries.keySet()) {
>                     long offset = ent.getOffset();
>                     int sectionID = ent.section == null ? 0 : s.getSectionID();
>                     relocTabStruct.createRelocationEntry(sectionID, (int) offset, syms.indexOf(ent.sym), (int) ent.t.toLong());
>                 }
>             }
>         }
> ```
> 
> here we have an `ent` in the innermost for where `ent.sym` is `null`. I will let you know if I find out more.

Yes, 

> > appears as if the relocation table is being written out before the symbol table. I'll poke around, and if there are any pointers to reproduce it in the CE version, that'd be great.
> 
> So far I found out that i can only reproduce it if build with `mx --strip-jars` so I think it only shows up if we have jar-files that have been stripped with proguard. Unfortunately I'm still not able to reproduce on CE.
> 
> The place where we issue occurs `PECoffRelocationTable.getNativeReloctab(PECoffRelocationTable.java:206` I see
> 
> ```java
>         for (PECoffSection s : getOwner().getPECoffSections()) {
>             Map<Entry, Entry> entries = (Map<Entry, Entry>) s.getRelocEntries();
>             if (entries != null) {
>                 for (Entry ent : entries.keySet()) {
>                     long offset = ent.getOffset();
>                     int sectionID = ent.section == null ? 0 : s.getSectionID();
>                     relocTabStruct.createRelocationEntry(sectionID, (int) offset, syms.indexOf(ent.sym), (int) ent.t.toLong());
>                 }
>             }
>         }
> ```
> 
> here we have an `ent` in the innermost for where `ent.sym` is `null`. I will let you know if I find out more.

Interesting - I would have thought that the stack trace would include HashTable if ent.sym was null.
Can you run with assertions enabled and see if line 152 in PECoffUserDefinedSection is triggered?
If so, the symbolname value would be quite valuable to solving this.

olpaw commented:
> Can you run with assertions enabled and see if line 152 in PECoffUserDefinedSection is triggered?
> If so, the symbolname value would be quite valuable to solving this.

Will do.

olpaw commented:
> > Can you run with assertions enabled and see if line 152 in PECoffUserDefinedSection is triggered?
> > If so, the symbolname value would be quite valuable to solving this.
> 
> Will do.

@stooke as you predicted, now I see the assertion:
```
[jvmcicompiler:11040]        write:     249.87 ms,  6.65 GB
Fatal error:java.lang.AssertionError
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
        at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
        at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:501)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:380)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:543)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:119)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:573)
Caused by: java.lang.AssertionError
        at com.oracle.objectfile.pecoff.PECoffUserDefinedSection.markRelocationSite(PECoffUserDefinedSection.java:152)
        at com.oracle.objectfile.BasicProgbitsSectionImpl.markRelocationSite(BasicProgbitsSectionImpl.java:118)
        at com.oracle.objectfile.pecoff.cv.CVLineRecord.computeHeader(CVLineRecord.java:95)
        at com.oracle.objectfile.pecoff.cv.CVLineRecord.computeContents(CVLineRecord.java:77)
        at com.oracle.objectfile.pecoff.cv.CVSymbolRecord.computeFullContents(CVSymbolRecord.java:57)
        at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.writeContent(CVSymbolSectionImpl.java:101)
        at com.oracle.objectfile.ObjectFile.debugContext(ObjectFile.java:1845)
        at com.oracle.objectfile.pecoff.cv.CVSectionImpl.getOrDecideContent(CVSectionImpl.java:84)
        at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.getOrDecideContent(CVSymbolSectionImpl.java:41)
        at com.oracle.objectfile.pecoff.PECoffUserDefinedSection.getOrDecideContent(PECoffUserDefinedSection.java:104)
        at com.oracle.objectfile.ObjectFile.bake(ObjectFile.java:1630)
        at com.oracle.objectfile.ObjectFile.write(ObjectFile.java:1266)
        at com.oracle.svm.hosted.image.NativeImage.lambda$write$0(NativeImage.java:168)
        at com.oracle.objectfile.ObjectFile.withDebugContext(ObjectFile.java:1826)
        at com.oracle.svm.hosted.image.NativeImage.write(NativeImage.java:167)
        at com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:409)
        at com.oracle.svm.hosted.image.SharedLibraryImageViaCC.write(SharedLibraryImageViaCC.java:56)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:701)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:497)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
[jvmcicompiler:11040]      [total]: 456,971.11 ms,  6.65 GB
```

olpaw commented:
> > > Can you run with assertions enabled and see if line 152 in PECoffUserDefinedSection is triggered?
> > > If so, the symbolname value would be quite valuable to solving this.
> > 
> > 
> > Will do.
> 
> @stooke as you predicted, now I see the assertion:
> 
> ```
> [jvmcicompiler:11040]        write:     249.87 ms,  6.65 GB
> Fatal error:java.lang.AssertionError
>         at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
>         at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
>         at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
>         at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
>         at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
>         at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
>         at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:501)
>         at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:380)
>         at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:543)
>         at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:119)
>         at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:573)
> Caused by: java.lang.AssertionError
>         at com.oracle.objectfile.pecoff.PECoffUserDefinedSection.markRelocationSite(PECoffUserDefinedSection.java:152)
>         at com.oracle.objectfile.BasicProgbitsSectionImpl.markRelocationSite(BasicProgbitsSectionImpl.java:118)
>         at com.oracle.objectfile.pecoff.cv.CVLineRecord.computeHeader(CVLineRecord.java:95)
>         at com.oracle.objectfile.pecoff.cv.CVLineRecord.computeContents(CVLineRecord.java:77)
>         at com.oracle.objectfile.pecoff.cv.CVSymbolRecord.computeFullContents(CVSymbolRecord.java:57)
>         at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.writeContent(CVSymbolSectionImpl.java:101)
>         at com.oracle.objectfile.ObjectFile.debugContext(ObjectFile.java:1845)
>         at com.oracle.objectfile.pecoff.cv.CVSectionImpl.getOrDecideContent(CVSectionImpl.java:84)
>         at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.getOrDecideContent(CVSymbolSectionImpl.java:41)
>         at com.oracle.objectfile.pecoff.PECoffUserDefinedSection.getOrDecideContent(PECoffUserDefinedSection.java:104)
>         at com.oracle.objectfile.ObjectFile.bake(ObjectFile.java:1630)
>         at com.oracle.objectfile.ObjectFile.write(ObjectFile.java:1266)
>         at com.oracle.svm.hosted.image.NativeImage.lambda$write$0(NativeImage.java:168)
>         at com.oracle.objectfile.ObjectFile.withDebugContext(ObjectFile.java:1826)
>         at com.oracle.svm.hosted.image.NativeImage.write(NativeImage.java:167)
>         at com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:409)
>         at com.oracle.svm.hosted.image.SharedLibraryImageViaCC.write(SharedLibraryImageViaCC.java:56)
>         at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:701)
>         at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:497)
>         at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
>         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
>         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
>         at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
>         at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
>         at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
> [jvmcicompiler:11040]      [total]: 456,971.11 ms,  6.65 GB
> ```

`PECoffSymtab.Entry ent = syms.getSymbol(symbolName);` return `null` for symbolName `"a_constructor_8c02267cc1acd96a5fb254a80e1944571dd3e21b"`

stooke commented:
> > > > Can you run with assertions enabled and see if line 152 in PECoffUserDefinedSection is triggered?
> > > > If so, the symbolname value would be quite valuable to solving this.
> > > 
> > > 
> > > Will do.
> > 
> > 
> > @stooke as you predicted, now I see the assertion:
> > ```
> > [jvmcicompiler:11040]        write:     249.87 ms,  6.65 GB
> > Fatal error:java.lang.AssertionError
> >         at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
> >         at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
> >         at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
> >         at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
> >         at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
> >         at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
> >         at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:501)
> >         at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:380)
> >         at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:543)
> >         at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:119)
> >         at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:573)
> > Caused by: java.lang.AssertionError
> >         at com.oracle.objectfile.pecoff.PECoffUserDefinedSection.markRelocationSite(PECoffUserDefinedSection.java:152)
> >         at com.oracle.objectfile.BasicProgbitsSectionImpl.markRelocationSite(BasicProgbitsSectionImpl.java:118)
> >         at com.oracle.objectfile.pecoff.cv.CVLineRecord.computeHeader(CVLineRecord.java:95)
> >         at com.oracle.objectfile.pecoff.cv.CVLineRecord.computeContents(CVLineRecord.java:77)
> >         at com.oracle.objectfile.pecoff.cv.CVSymbolRecord.computeFullContents(CVSymbolRecord.java:57)
> >         at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.writeContent(CVSymbolSectionImpl.java:101)
> >         at com.oracle.objectfile.ObjectFile.debugContext(ObjectFile.java:1845)
> >         at com.oracle.objectfile.pecoff.cv.CVSectionImpl.getOrDecideContent(CVSectionImpl.java:84)
> >         at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.getOrDecideContent(CVSymbolSectionImpl.java:41)
> >         at com.oracle.objectfile.pecoff.PECoffUserDefinedSection.getOrDecideContent(PECoffUserDefinedSection.java:104)
> >         at com.oracle.objectfile.ObjectFile.bake(ObjectFile.java:1630)
> >         at com.oracle.objectfile.ObjectFile.write(ObjectFile.java:1266)
> >         at com.oracle.svm.hosted.image.NativeImage.lambda$write$0(NativeImage.java:168)
> >         at com.oracle.objectfile.ObjectFile.withDebugContext(ObjectFile.java:1826)
> >         at com.oracle.svm.hosted.image.NativeImage.write(NativeImage.java:167)
> >         at com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:409)
> >         at com.oracle.svm.hosted.image.SharedLibraryImageViaCC.write(SharedLibraryImageViaCC.java:56)
> >         at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:701)
> >         at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:497)
> >         at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
> >         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
> >         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
> >         at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
> >         at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
> >         at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
> > [jvmcicompiler:11040]      [total]: 456,971.11 ms,  6.65 GB
> > ```
> 
> @stooke `PECoffSymtab.Entry ent = syms.getSymbol(symbolName);` returns `null` for **symbolName** `"a_constructor_8c02267cc1acd96a5fb254a80e1944571dd3e21b"`

Thanks, I'll do some poking around and get back.

stooke commented:
> > > @stooke as you predicted, now I see the assertion:
> > > ```
......
> > > Caused by: java.lang.AssertionError
> > >         at com.oracle.objectfile.pecoff.PECoffUserDefinedSection.markRelocationSite(PECoffUserDefinedSection.java:152)
> > >         at com.oracle.objectfile.BasicProgbitsSectionImpl.markRelocationSite(BasicProgbitsSectionImpl.java:118)
> > >         at com.oracle.objectfile.pecoff.cv.CVLineRecord.computeHeader(CVLineRecord.java:95)
> > >         at com.oracle.objectfile.pecoff.cv.CVLineRecord.computeContents(CVLineRecord.java:77)
> > >         at com.oracle.objectfile.pecoff.cv.CVSymbolRecord.computeFullContents(CVSymbolRecord.java:57)
> > >         at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.writeContent(CVSymbolSectionImpl.java:101)
> > >         at com.oracle.objectfile.ObjectFile.debugContext(ObjectFile.java:1845)
> > >         at com.oracle.objectfile.pecoff.cv.CVSectionImpl.getOrDecideContent(CVSectionImpl.java:84)
> > >         at com.oracle.objectfile.pecoff.cv.CVSymbolSectionImpl.getOrDecideContent(CVSymbolSectionImpl.java:41)
> > >         at com.oracle.objectfile.pecoff.PECoffUserDefinedSection.getOrDecideContent(PECoffUserDefinedSection.java:104)
> > >         at com.oracle.objectfile.ObjectFile.bake(ObjectFile.java:1630)
> > >         at com.oracle.objectfile.ObjectFile.write(ObjectFile.java:1266)
> > >         at com.oracle.svm.hosted.image.NativeImage.lambda$write$0(NativeImage.java:168)
> > >         at com.oracle.objectfile.ObjectFile.withDebugContext(ObjectFile.java:1826)
> > >         at com.oracle.svm.hosted.image.NativeImage.write(NativeImage.java:167)
> > >         at com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:409)
> > >         at com.oracle.svm.hosted.image.SharedLibraryImageViaCC.write(SharedLibraryImageViaCC.java:56)
> > >         at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:701)
> > >         at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:497)
> > >         at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
> > >         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
> > >         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
> > >         at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
> > >         at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
> > >         at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
> > > [jvmcicompiler:11040]      [total]: 456,971.11 ms,  6.65 GB
> > > ```
> > 
> > 
> > @stooke `PECoffSymtab.Entry ent = syms.getSymbol(symbolName);` returns `null` for **symbolName** `"a_constructor_8c02267cc1acd96a5fb254a80e1944571dd3e21b"`
> 
> What appears to be happening is the Windows debug code has encountered a function (PrimaryEntry) that has no global symbol table entry.

I propose a temporary work around - at the place where the assertion is fired, the code tests and if the entry is not found, emits a warning message (and a dummy relocation entry if required to keep the same size, but I don't think it is) and avoid adding a bad relocation entry in the first place.

If this is acceptable (for now) I can prepare a patch.  This workaround is low risk and low impact, but not the final solution.

olpaw commented:
> If this is acceptable (for now) I can prepare a patch. This workaround is low risk and low impact, but not the final solution.

Makes sense so that @zakkak can get his PR merged and continue with the implementation of debuginfo for inlined methods.
Please commit the proposed workaround to this PR.

stooke commented:
> > If this is acceptable (for now) I can prepare a patch. This workaround is low risk and low impact, but not the final solution.
> 
> Makes sense - so that @zakkak can get his PR merged and continue with the implementation of debuginfo for inlined methods.
> @stooke please commit the proposed workaround to this PR.

I've had some issues opening PRs against @zakkak 's repo - I'm chasing them down but have sent him the patch.  Actually I made 2 patches:
- The first patch is more 'correct' in that is changes am interface function return value (and many implementations thereof).  The return value is never used, at least in the CE version of Graal.
- The second, much smaller patch, returns a dummy value that is ignored everywhere.

I mention this because my time zone is fairly late compared to @zakkak so I thought this lets the conversation start early.

olpaw commented:
@zakkak please ping me on slack today when you have the changes from @stooke on your branch.

zakkak commented:
Hi @olpaw, I have integrated @stooke's patches and squashed the revert and fixup commits as mentioned in https://github.com/oracle/graal/pull/3419#issuecomment-858640166 .

The new changes are in https://github.com/oracle/graal/pull/3419/commits/7d01a04433c6f0475cbdc5d6d7fdd567cb6ae326 (Simon's minimal patch to work around the issue) and https://github.com/oracle/graal/pull/3419/commits/2676016cdc7e66e3c8f2ab5755e0b19bc034edaa (Simon's larger patch with some refactoring in place).

If things look good I suggest squashing the two commits in one.

olpaw commented:
Hi @zakkak with this patches from @stooke even building fails nows:
```
[16:10:04] Compiling com.oracle.graal.pointsto with javac-daemon(JDK 1.8)... [dependency GRAAL_PROCESSOR updated]
/b/b/e/main/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:347: error: cannot find symbol
            int comparisonResult = methodEntry.compareTo(methodName, paramSignature, returnTypeName);
                                                         ^
  symbol:   variable methodName
  location: class ClassEntry
/b/b/e/main/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:347: error: cannot find symbol
            int comparisonResult = methodEntry.compareTo(methodName, paramSignature, returnTypeName);
                                                                     ^
  symbol:   variable paramSignature
  location: class ClassEntry
/b/b/e/main/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:347: error: cannot find symbol
            int comparisonResult = methodEntry.compareTo(methodName, paramSignature, returnTypeName);
                                                                                     ^
  symbol:   variable returnTypeName
  location: class ClassEntry
```

zakkak commented:
> Hi @zakkak with the patches from @stooke even building fails nows:

Sorry @olpaw , that's my mistake, some changes that should have been reverted got away with the rebase. It should be fixed now.

olpaw commented:
> Sorry @olpaw , that's my mistake, some changes that should have been reverted got away with the rebase. It should be fixed now.

Alright, rerunning CI gates now ...

olpaw commented:
> Alright, rerunning CI gates now ...

@zakkak this time the CI passed just fine. ... except for one truffle-ruby gate that fails which clearly has nothing to do with your changes. Once we get the truffle-ruby gate fixed your PR will be in the merge queue. There is nothing more to do on your side. With a bit of luck the PR should be on master by then of this week.

zakkak commented:
Nice, thank you Paul!

olpaw commented:
Now in the merge queue (on 3rd place)
